File: ExelBidDemoObjC/ExelBidDemoObjC/MPartners/MPartnersNativeController.m
#import "MPartnersNativeController.h"
#import "CustomNativeAdView.h"
#import "DesignSystem.h"
#import "StatusBadge.h"
#import "AdUnitIds.h"
@import ExelBidSDK;

/// MPartners native ad. Reuses `CustomNativeAdView`.
/// Note: MPartners native responses do not include `imp50tracker` /
/// `imp100tracker`, so the 50% / 100% impression callbacks never fire.
@interface MPartnersNativeController ()
@property (nonatomic, strong) EBStatusBadge *badge;
@property (nonatomic, strong) EBLogView *logView;
@property (nonatomic, strong) UIView *adContainer;
@property (nonatomic, strong, nullable) EBNativeAd *loadedAd;
@property (nonatomic, strong, nullable) EBMPartnersNativeAdLoader *loader;
@end

@implementation MPartnersNativeController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.view.backgroundColor = UIColor.systemGroupedBackgroundColor;
    self.title = @"MPartners Native";

    UIScrollView *scroll = [[UIScrollView alloc] init];
    scroll.translatesAutoresizingMaskIntoConstraints = NO;
    scroll.alwaysBounceVertical = YES;
    [self.view addSubview:scroll];

    UIStackView *stack = [[UIStackView alloc] init];
    stack.axis = UILayoutConstraintAxisVertical;
    stack.spacing = EBSpacingL;
    stack.translatesAutoresizingMaskIntoConstraints = NO;
    [scroll addSubview:stack];

    EBCardView *status = [[EBCardView alloc] init];
    [status addArranged:[EBSectionLabel make:@"Status"]];
    self.badge = [[EBStatusBadge alloc] initWithFrame:CGRectZero];
    [status addArranged:self.badge];
    [stack addArrangedSubview:status];

    EBCardView *actions = [[EBCardView alloc] init];
    [actions addArranged:[EBSectionLabel make:@"Action"]];
    [actions addArranged:[EBPrimaryButton filled:@"Load MPartners native ad" target:self action:@selector(handleLoad)]];
    [stack addArrangedSubview:actions];

    EBCardView *creative = [[EBCardView alloc] init];
    [creative addArranged:[EBSectionLabel make:@"Creative"]];
    self.adContainer = [[UIView alloc] init];
    self.adContainer.translatesAutoresizingMaskIntoConstraints = NO;
    [creative addArranged:self.adContainer];
    [stack addArrangedSubview:creative];

    EBCardView *note = [[EBCardView alloc] init];
    [note addArranged:[EBSectionLabel make:@"Note"]];
    [note addArranged:[EBTypography footnote:@"MPartners 응답에는 imp50tracker / imp100tracker가 포함되지 않습니다. 50% / 100% 노출 콜백은 발생하지 않으며 attach 시점의 단일 impression만 처리됩니다."]];
    [stack addArrangedSubview:note];

    EBCardView *log = [[EBCardView alloc] init];
    [log addArranged:[EBSectionLabel make:@"Lifecycle log"]];
    self.logView = [[EBLogView alloc] initWithFrame:CGRectZero];
    [log addArranged:self.logView];
    [stack addArrangedSubview:log];

    [NSLayoutConstraint activateConstraints:@[
        [scroll.topAnchor      constraintEqualToAnchor:self.view.safeAreaLayoutGuide.topAnchor],
        [scroll.bottomAnchor   constraintEqualToAnchor:self.view.safeAreaLayoutGuide.bottomAnchor],
        [scroll.leadingAnchor  constraintEqualToAnchor:self.view.leadingAnchor],
        [scroll.trailingAnchor constraintEqualToAnchor:self.view.trailingAnchor],
        [stack.topAnchor      constraintEqualToAnchor:scroll.contentLayoutGuide.topAnchor constant:EBSpacingL],
        [stack.bottomAnchor   constraintEqualToAnchor:scroll.contentLayoutGuide.bottomAnchor constant:-EBSpacingXXL],
        [stack.leadingAnchor  constraintEqualToAnchor:scroll.frameLayoutGuide.leadingAnchor constant:EBSpacingL],
        [stack.trailingAnchor constraintEqualToAnchor:scroll.frameLayoutGuide.trailingAnchor constant:-EBSpacingL]
    ]];

    [self.badge setState:EBStatusBadgeStateIdle text:@"Tap to load"];
}

- (void)handleLoad {
    [self.badge setState:EBStatusBadgeStateLoading text:@"loading…"];
    [self.logView append:@"load requested"];

    EBMPartnersNativeAdLoader *loader =
        [[EBMPartnersNativeAdLoader alloc] initWithAdUnitId:AdUnitIdsMPartners.native];
    loader.desiredAssetsArray = @[
        @(EBNativeAssetTitle), @(EBNativeAssetMain), @(EBNativeAssetIcon),
        @(EBNativeAssetCtatext), @(EBNativeAssetDesc)
    ];
    self.loader = loader;

    __weak typeof(self) weakSelf = self;
    [loader loadWithCompletion:^(EBNativeAd * _Nullable ad, NSError * _Nullable error) {
        typeof(self) strongSelf = weakSelf;
        if (!strongSelf) { return; }
        if (!ad) {
            [strongSelf setState:EBStatusBadgeStateFailed
                            text:[NSString stringWithFormat:@"failed: %@", error.localizedDescription]];
            return;
        }
        [strongSelf.loadedAd detach];
        for (UIView *sub in strongSelf.adContainer.subviews) { [sub removeFromSuperview]; }

        CustomNativeAdView *adView = [[CustomNativeAdView alloc] init];
        adView.translatesAutoresizingMaskIntoConstraints = NO;
        [strongSelf.adContainer addSubview:adView];
        [NSLayoutConstraint activateConstraints:@[
            [adView.leadingAnchor  constraintEqualToAnchor:strongSelf.adContainer.leadingAnchor],
            [adView.trailingAnchor constraintEqualToAnchor:strongSelf.adContainer.trailingAnchor],
            [adView.topAnchor      constraintEqualToAnchor:strongSelf.adContainer.topAnchor],
            [adView.bottomAnchor   constraintEqualToAnchor:strongSelf.adContainer.bottomAnchor]
        ]];

        ad.presenterProvider = ^UIViewController * _Nullable { return weakSelf; };
        ad.onImpression  = ^{ [weakSelf setState:EBStatusBadgeStateImpression text:@"impression"]; };
        ad.onClick       = ^{ [weakSelf setState:EBStatusBadgeStateClicked    text:@"clicked"]; };
        ad.onLeaveApp    = ^{ [weakSelf setState:EBStatusBadgeStateClicked    text:@"left app"]; };
        ad.onClickFinish = ^{ [weakSelf setState:EBStatusBadgeStateReady      text:@"click finished"]; };

        [ad attachTo:adView];
        strongSelf.loadedAd = ad;
        [strongSelf setState:EBStatusBadgeStateReady text:@"attached"];
    }];
}

- (void)setState:(EBStatusBadgeState)state text:(NSString *)text {
    [self.badge setState:state text:text];
    [self.logView append:text];
}

@end

File: ExelBidDemo/ExelBidDemo/MPartners/MPartnersNativeController.swift
import UIKit
import ExelBidSDK

/// MPartners native ad. Reuses `CustomNativeAdView` (the standard
/// `EBNativeAdRendering` view) — the SDK returns the same `EBNativeAd` type
/// regardless of which backend served it.
///
/// Note: MPartners native responses do not include `imp50tracker` /
/// `imp100tracker`, so the 50% / 100% impression callbacks never fire.
/// The single `onImpression` (immediate, on attach) is the only one used.
final class MPartnersNativeController: UIViewController {

    private let badge = StatusBadge()
    private let logView = LogView()
    private let options = TestOptionsPanel()
    private let adContainer = UIView()
    private var loadedAd: EBNativeAd?

    override func viewDidLoad() {
        super.viewDidLoad()
        view.backgroundColor = .systemGroupedBackground
        title = "MPartners Native"

        let scroll = UIScrollView()
        scroll.translatesAutoresizingMaskIntoConstraints = false
        scroll.alwaysBounceVertical = true
        view.addSubview(scroll)

        let stack = UIStackView()
        stack.axis = .vertical
        stack.spacing = Spacing.l
        stack.translatesAutoresizingMaskIntoConstraints = false
        scroll.addSubview(stack)

        let status = CardView()
        status.addArranged(SectionLabel.make("Status"))
        status.addArranged(badge)
        stack.addArrangedSubview(status)

        stack.addArrangedSubview(options)

        let actions = CardView()
        actions.addArranged(SectionLabel.make("Action"))
        actions.addArranged(PrimaryButton.filled("Load MPartners native ad", target: self, action: #selector(handleLoad)))
        stack.addArrangedSubview(actions)

        let creative = CardView()
        creative.addArranged(SectionLabel.make("Creative"))
        adContainer.translatesAutoresizingMaskIntoConstraints = false
        creative.addArranged(adContainer)
        stack.addArrangedSubview(creative)

        let note = CardView()
        note.addArranged(SectionLabel.make("Note"))
        note.addArranged(Typography.footnote(
            "MPartners 네이티브 광고는 50% / 100% 가시성 노출 측정을 지원하지 않습니다. " +
            "광고 뷰가 화면에 표시되는 시점의 노출 이벤트만 호출됩니다."
        ))
        stack.addArrangedSubview(note)

        let log = CardView()
        log.addArranged(SectionLabel.make("Lifecycle log"))
        log.addArranged(logView)
        stack.addArrangedSubview(log)

        NSLayoutConstraint.activate([
            scroll.topAnchor.constraint(equalTo: view.safeAreaLayoutGuide.topAnchor),
            scroll.bottomAnchor.constraint(equalTo: view.safeAreaLayoutGuide.bottomAnchor),
            scroll.leadingAnchor.constraint(equalTo: view.leadingAnchor),
            scroll.trailingAnchor.constraint(equalTo: view.trailingAnchor),

            stack.topAnchor.constraint(equalTo: scroll.contentLayoutGuide.topAnchor, constant: Spacing.l),
            stack.bottomAnchor.constraint(equalTo: scroll.contentLayoutGuide.bottomAnchor, constant: -Spacing.xxl),
            stack.leadingAnchor.constraint(equalTo: scroll.frameLayoutGuide.leadingAnchor, constant: Spacing.l),
            stack.trailingAnchor.constraint(equalTo: scroll.frameLayoutGuide.trailingAnchor, constant: -Spacing.l)
        ])

        badge.set(.idle, text: "Tap to load")
    }

    @objc private func handleLoad() {
        badge.set(.loading, text: "loading…")
        logView.append("load requested")
        Task { [weak self] in await self?.loadAd() }
    }

    @MainActor
    private func loadAd() async {
        let loader = EBMPartnersNativeAdLoader(adUnitId: AdUnitIds.MPartners.native)
        options.apply(to: loader.options)
        loader.desiredAssets = [.title, .main, .icon, .ctatext, .desc]

        do {
            let ad = try await loader.load()
            loadedAd?.detach()
            adContainer.subviews.forEach { $0.removeFromSuperview() }

            let adView = CustomNativeAdView()
            adView.translatesAutoresizingMaskIntoConstraints = false
            adContainer.addSubview(adView)
            NSLayoutConstraint.activate([
                adView.leadingAnchor.constraint(equalTo: adContainer.leadingAnchor),
                adView.trailingAnchor.constraint(equalTo: adContainer.trailingAnchor),
                adView.topAnchor.constraint(equalTo: adContainer.topAnchor),
                adView.bottomAnchor.constraint(equalTo: adContainer.bottomAnchor)
            ])

            ad.presenterProvider = { [weak self] in self }
            ad.onImpression  = { [weak self] in self?.set(.impression, "impression") }
            ad.onClick       = { [weak self] in self?.set(.clicked,    "clicked") }
            ad.onLeaveApp    = { [weak self] in self?.set(.clicked,    "left app") }
            ad.onClickFinish = { [weak self] in self?.set(.ready,      "click finished") }

            ad.attach(to: adView)
            self.loadedAd = ad
            set(.ready, "attached")
        } catch {
            set(.failed, "failed: \(error.localizedDescription)")
        }
    }

    private func set(_ state: StatusBadge.State, _ text: String) {
        badge.set(state, text: text)
        logView.append(text)
    }
}
